pkg/upgrade/control_plane.go

```diff
@@ -279,67 +328,124 @@ func (u *ControlPlaneUpgrader) updateMachines(machines *clusterapiv1alpha1.Machi
+		// let's set the new name to controlplane-<index>-<timestamp>
+		nameParts := strings.Split(machine.Name, "-")
+		if len(nameParts) < 2 {
+			return errors.Errorf("machine name %q does not match expected format <name>-<index> or <name>-<index>-<timestamp>", machine.Name)
 		}
+		name := fmt.Sprintf("%s-%s-%d", nameParts[0], nameParts[1], time.Now().Unix())
+		// TODO: generate the name based off each respective object
 
-		oldHostName := hostnameForNode(oldNode)
-
-		newMachine, node, err := machineCreator.NewMachine(&machine)
+		err = u.updateInfrastructureReference(name, &machine)
 		if err != nil {
 			return err
 		}
-		nodeHostname := hostnameForNode(node)
-
-		// This used to happen when a new machine was created as a side effect. Must still update the mapping.
-		if err := u.UpdateProviderIDsToNodes(); err != nil {
-			return err
-		}
 
-		// delete old etcd member
-		err = u.deleteEtcdMember(time.Minute*1, nodeHostname, u.oldNodeToEtcdMember[oldHostName])
+		err = u.updateBootstrapConfig(name, &machine)
 		if err != nil {
-			return errors.Wrapf(err, "unable to delete old etcd member %s", u.oldNodeToEtcdMember[oldHostName])
-		}
-
-		if err := u.deleteMachine(&machine); err != nil {
 			return err
 		}
 
-		if err := u.applyAnnotation(newMachine); err != nil {
+		if err := u.updateMachine(name, machine, machineCreator); err != nil {
 			return err
 		}
 	}
+
 	return nil
 }
 
-func (u *ControlPlaneUpgrader) applyAnnotation(m *clusterapiv1alpha1.Machine) error {
-	original, err := json.Marshal(m)
+func (u *ControlPlaneUpgrader) updateBootstrapConfig(name string, machine *clusterv1.Machine) error {
+	// copy node registration
+	bootstrap := &bootstrapv1.KubeadmConfig{}
+	if machine.Spec.Bootstrap.ConfigRef.Namespace == "" {
+		machine.Spec.Bootstrap.ConfigRef.Namespace = u.clusterNamespace
+	}
+	err := u.managerClusterClient.Get(context.TODO(), ctrlclient.ObjectKey{Name: machine.Spec.Bootstrap.ConfigRef.Name, Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace}, bootstrap)
 	if err != nil {
 		return errors.WithStack(err)
 	}
-	if m.Annotations == nil {
-		m.Annotations = map[string]string{}
+
+	// modify bootstrap config
+	bootstrap.SetName(name)
+	bootstrap.SetResourceVersion("")
+	bootstrap.SetOwnerReferences(nil)
+
+	// find node registration
+	nodeRegistration := kubeadmv1beta1.NodeRegistrationOptions{}
+	if bootstrap.Spec.InitConfiguration != nil {
+		nodeRegistration = bootstrap.Spec.InitConfiguration.NodeRegistration
+	} else if bootstrap.Spec.JoinConfiguration != nil {
+		nodeRegistration = bootstrap.Spec.JoinConfiguration.NodeRegistration
 	}
-	m.Annotations[UpgradeIDAnnotationKey] = u.upgradeID
-	updated, err := json.Marshal(m)
+	if bootstrap.Spec.JoinConfiguration == nil {
+		bootstrap.Spec.JoinConfiguration = &kubeadmv1beta1.JoinConfiguration{
+			ControlPlane: &kubeadmv1beta1.JoinControlPlane{},
+		}
+	}
+	bootstrap.Spec.JoinConfiguration.NodeRegistration = nodeRegistration
+
+	// clear init configuration
```

Comment:
Should we expand on why this is needed?

Reply:
Previously I noticed that when you have both the init configuration and the join configuration present for a control plane upgrade, kubeadm will use the init configuration instead of the join configuration. Also during upgrades, you will never be initializing a new node. It will always be joining an existing control plane.

Reply:
Will add a comment.